### PR TITLE
overlord: execute snapshot cleanup in task

### PR DESCRIPTION
**Pull request by mardy**
This is an alternative to https://github.com/snapcore/snapd/pull/11321

Here we do not alter the logic for conflict detection, but simply move the shanpshot cleanup as part of the corresponding task, and not as a cleanup handler.

### CODE REVIEW

File: overlord/snapshotstate/snapshotmgr.go [hunk 1]
```diff
@@ -371,33 +372,6 @@ func undoRestore(task *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
-func cleanupRestore(task *state.Task, _ *tomb.Tomb) error {
```
Comment (MiguelPires) [inline overlord/snapshotstate/snapshotmgr.go:374]: The reference to this needs to be removed from `export_test.go`

Comment (mardy) [reply overlord/snapshotstate/snapshotmgr.go:374]: I force-pushed a new version, which should actually build. Given that I had to remove a unit test, I updated another one to ensure that `Cleanup()` is being called.

File: overlord/snapshotstate/snapshotmgr.go [hunk 2]
```diff
@@ -339,6 +338,8 @@ func doRestore(task *state.Task, tomb *tomb.Tomb) error {
 	restoreState.Config = oldCfg
 	task.Set("restore-state", restoreState)
 
+	backendCleanup(restoreState)
```
Comment (pedronis) [inline overlord/snapshotstate/snapshotmgr.go:341]: this is all confusing, but I suspect we can't simply do this because it breaks undos?

Comment (pedronis) [reply overlord/snapshotstate/snapshotmgr.go:341]: as undoRestore will want to use restoreState. I don't know if we have relevant tests or not for this :/

Comment (mardy) [reply overlord/snapshotstate/snapshotmgr.go:341]: > the problem is that as designed a restore change can have many restore tasks, some might succeed but if one fails everything is undone, but if we do this, we lost the state of the one that succeeded when we try to undo all of them

But does this really matter here? I guess that a restore change can have many restore tasks *on different snaps*, right? In our specific case, the cleanup handler goal is to get rid of the temporary files created during the restore operation, so I don't think it matters whether it's executed when all tasks are done, or right after the corresponding task is done: the end result will be the same.

I might need to add a call to `backendCleanup()` also on the undo task (I've yet to verify if it's needed), but other than that I don't see an issue with this. Or can you make an example of when things could go wrong?

Comment (pedronis) [reply overlord/snapshotstate/snapshotmgr.go:341]: the issue is that in many snaps restore where one fails,  with this code we end up calling Revert on restoreState that has been Cleanup for the snaps that succeeded, and that can/will fail

File: overlord/snapshotstate/snapshotmgr.go [hunk 3]
```diff
@@ -371,6 +372,23 @@ func undoRestore(task *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+func doTaskCleanup(task *state.Task, tomb *tomb.Tomb) error {
+	return cleanupRestore(task, tomb)
+}
+
+func doCleanup(task *state.Task, tomb *tomb.Tomb) error {
+	for _, t := range task.WaitTasks() {
```
Comment (MiguelPires) [inline overlord/snapshotstate/snapshotmgr.go:380]: This is accessed without acquiring the state lock (there are some unit test failures)

File: overlord/snapshotstate/snapshotmgr.go [hunk 4]
```diff
@@ -371,6 +372,23 @@ func undoRestore(task *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+func doTaskCleanup(task *state.Task, tomb *tomb.Tomb) error {
```
Comment (MiguelPires) [inline overlord/snapshotstate/snapshotmgr.go:375]: Why not just use `cleanupRestore()` directly?

File: overlord/snapshotstate/snapshotmgr.go (hunk 3 above)
Comment (mardy) [reply overlord/snapshotstate/snapshotmgr.go]: Thanks, you were correct indeed!

File: overlord/snapshotstate/snapshotmgr.go (hunk 4 above)
Comment (mardy) [reply overlord/snapshotstate/snapshotmgr.go]: I thought I would need to do some other operations before or after it, but it turns out that using `cleaupRestore()` directly is enough. I've removed the wrapper now.

File: overlord/snapshotstate/snapshotstate.go [hunk 5]
```diff
@@ -510,6 +510,13 @@ func Restore(st *state.State, setID uint64, snapNames []string, users []string)
 		ts.AddTask(task)
 	}
 
+	if len(summaries) > 0 {
+		desc := fmt.Sprintf("Cleanup of restore snapshot set #%d", setID)
```
Comment (pedronis) [inline overlord/snapshotstate/snapshotstate.go:514]: "Cleanup after restore from snapshot set #%d"

Comment (pedronis) [inline overlord/snapshotstate/snapshotstate.go:514]: this merits a comment:

// take care of cleaning up all restore working state if all the restore tasks succeeded,
// if they didn't the undo logic will take care of this

File: overlord/snapshotstate/snapshotmgr.go [hunk 6]
```diff
@@ -371,6 +371,23 @@ func undoRestore(task *state.Task, _ *tomb.Tomb) error {
...
 }
 
+func doCleanup(task *state.Task, tomb *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	restoreTasks := task.WaitTasks()
+	st.Unlock()
+	for _, t := range restoreTasks {
+		if err := cleanupRestore(t, tomb); err != nil {
+			logger.Noticef("Cleanup for tag %s failed", task.ID())
```
Comment (pedronis) [inline overlord/snapshotstate/snapshotmgr.go:381]: "Cleanup of restore task %s failed"

File: overlord/snapshotstate/snapshotmgr.go [hunk 7]
```diff
@@ -73,7 +73,7 @@ func Manager(st *state.State, runner *state.TaskRunner) *SnapshotManager {
 	runner.AddHandler("forget-snapshot", doForget, nil)
 	runner.AddHandler("check-snapshot", doCheck, nil)
 	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
-	runner.AddCleanup("restore-snapshot", cleanupRestore)
+	runner.AddHandler("cleanup-restore-snapshot", doCleanup, nil)
```
Comment (pedronis) [inline overlord/snapshotstate/snapshotmgr.go:76]: this takes care of multiple snapshots so it's not clear if it should follow the naming of the other tasks, so maybe:

cleanup-after-restore and doCleanupAfterRestore

?

**Review by pedronis — COMMENTED** [review]
lgtm, some naming and message considerations

**Review by MiguelPires — APPROVED** [review]
LGTM, thanks

File: overlord/snapshotstate/snapshotmgr.go (hunk 6 above)
Comment (pedronis) [reply overlord/snapshotstate/snapshotmgr.go:381]: this still needs addressing afaict

Comment (mardy) [reply overlord/snapshotstate/snapshotmgr.go]: oops, done!

**Pull request merged**